src/model_theory/basic.lean

```diff
@@ -0,0 +1,328 @@
+  cases g,
+  simp only,
+  ext x,
+  exact function.funext_iff.1 h x,
+end
+
+@[ext]
+lemma ext ⦃f g : M ↪[L] N⦄ (h : ∀ x, f x = g x) : f = g :=
+coe_inj (funext h)
+
+lemma ext_iff {f g : M ↪[L] N} : f = g ↔ ∀ x, f x = g x :=
+⟨λ h x, h ▸ rfl, λ h, ext h⟩
+
+@[simp] lemma map_fun (φ : M ↪[L] N) {n : ℕ} (f : L.functions n) (x : fin n → M) :
+  φ (fun_map f x) = fun_map f (φ ∘ x) := φ.map_fun' f x
+
+@[simp] lemma map_rel (φ : M ↪[L] N) {n : ℕ} (r : L.relations n) (x : fin n → M) :
+  rel_map r (φ ∘ x) ↔ rel_map r x := φ.map_rel' r x
+
+lemma injective (f : M ↪[L] N) : function.injective f := f.to_embedding.injective
+
+variables (L) (M)
+/-- The identity embedding from a structure to itself-/
+def id : M ↪[L] M :=
+{ to_embedding := function.embedding.refl M,
+  map_fun' := λ _ _ _, rfl,
+  map_rel' := λ _ _ _, iff.refl _ }
+
+variables {L} {M}
+
+instance : inhabited (M ↪[L] M) := ⟨id L M⟩
+
+@[simp] lemma id_apply (x : M) :
+  id L M x = x := rfl
+
+variables {P : Type u} [L.Structure P] {Q : Type u} [L.Structure Q]
+
+/-- Composition of first-order embeddings -/
+def comp (hnp : N ↪[L] P) (hmn : M ↪[L] N) : M ↪[L] P :=
+{ to_fun := hnp ∘ hmn,
+  inj' := hnp.injective.comp hmn.injective,
+  map_fun' := λ n f, by simp,
+  map_rel' := λ _ _ _, by rw [map_rel, map_rel], }
+
+--infixr  ` ∘ `:80      := comp
+
+@[simp] lemma comp_apply (g : N ↪[L] P) (f : M ↪[L] N) (x : M) :
+  g.comp f x = g (f x) := rfl
+
+/-- Composition of first-order embeddings is associative. -/
+lemma comp_assoc (f : M ↪[L] N) (g : N ↪[L] P) (h : P ↪[L] Q) :
+  (h.comp g).comp f = h.comp (g.comp f) := rfl
+
+end embedding
+
+namespace equiv
+
+/-- The inverse of a first-order equivalence is a first-order equivalence. -/
+def symm (f : M ≃[L] N) : N ≃[L] M :=
+{ map_fun' := λ n f' x, begin
+    simp only [equiv.to_fun_as_coe],
+    rw [equiv.symm_apply_eq],
+    refine eq.trans _ (f.map_fun' f' (f.to_equiv.symm ∘ x)).symm,
+    rw [← function.comp.assoc, equiv.to_fun_as_coe, equiv.self_comp_symm, function.comp.left_id]
+  end,
+  map_rel' := λ n r x, begin
+    simp only [equiv.to_fun_as_coe],
+    refine (f.map_rel' r (f.to_equiv.symm ∘ x)).symm.trans _,
+    rw [← function.comp.assoc, equiv.to_fun_as_coe, equiv.self_comp_symm, function.comp.left_id]
+  end,
+  .. f.to_equiv.symm }
+
+/-- A first-order equivalence is also a first-order embedding. -/
+def to_embedding (f : M ≃[L] N) : M ↪[L] N :=
+{ to_embedding := f.to_equiv.to_embedding,
+  map_fun' := f.map_fun',
+  map_rel' := f.map_rel' }
+
+/-- A first-order equivalence is also a first-order embedding. -/
+def to_hom (f : M ≃[L] N) : M →[L] N :=
+{ to_fun := f.to_equiv,
+  map_fun' := f.map_fun',
+  map_rel' := λ n r x, (f.map_rel' r x).2 }
+
+@[simp] lemma to_embedding_to_hom (f : M ≃[L] N) : f.to_embedding.to_hom = f.to_hom := rfl
+
+@[simps] instance has_coe_to_fun : has_coe_to_fun (M ≃[L] N) :=
+⟨(λ _, M → N), λ f, f.to_embedding⟩
+
+@[simp]
+lemma coe_eq_to_hom {f : M ≃[L] N} : (f : M → N) = f.to_hom := rfl
```

Comment:
Usually we would have such a simp lemma run the other way: prefer to use coercions!